docs/book/stacks-and-components/component-guide/writing-custom-components.md

```diff
@@ -0,0 +1,239 @@
+---
+description: How to extend ZenML with plug-in flavor components.
+---
+
+# Develop a Custom Component Flavor in ZenML
+
+ZenML is built on the philosophy of composability and reusability. It allows users to develop custom component flavors for each type of pipeline component. This guide will help you understand what a flavor is, and how you can develop and register your own custom flavor in ZenML.
+
+## Understanding Component Flavors
+
+In ZenML, a component type is a broad category that defines the functionality of a pipeline component. Each type can have multiple flavors, which are specific implementations of the component type. For instance, the type `orchestrator` can have flavors like `local`, `kubernetes`, etc. Each flavor is a unique implementation of the orchestrator functionality.
+
+In order to create a custom component, one must create:
+
+1. A custom flavor class (inheriting from `zenml.stack.Flavor` or any of its child classes)
+2. An implementation class (inheriting from `zenml.stack.StackComponent` or any of its child classes)
+
+## Writing a Custom Flavor Class
+
+To develop a custom flavor, you need to create a class that inherits from the `BaseFlavor` class and implement the abstract properties. Here is a simplified version of the `BaseFlavor` class. Read the docstring of each function to understand its use:
+
+```python
+from abc import ABC, abstractmethod
+from typing import Type
+
+from zenml.enums import StackComponentType
+from zenml.stack import StackComponent, StackComponentConfig, Flavor
+
+
+class BaseFlavorConfig(StackComponentConfig):
+    """Base class for all ZenML flavor configurations.
+    
+    This is a pydantic class, so pydantic supported items can be
+    defined here.
+    """
+
+class BaseFlavor(Flavor, ABC):
+    """Base class for all ZenML flavors."""
+
+    @property
+    def name(self) -> str:
+        """The flavor name.
+
+        Returns:
+            The flavor name.
+        """
+
+    @property
+    def docs_url(self) -> Optional[str]:
+        """A url to point at docs explaining this flavor.
+
+        Returns:
+            A flavor docs url.
+        """
+        return None
+
+    @property
+    def sdk_docs_url(self) -> Optional[str]:
+        """A url to point at SDK docs explaining this flavor.
+
+        Returns:
+            A flavor SDK docs url.
+        """
+        return None
```

Comment:
Are these relevant for user created flavors? If not we can leave them out in the docs

Reply:
I would say we can even leave out this and the code sample below out completely? They'll get outdated anyway, and people implementing it will import/inspect the file in their IDE anyway and won't get much value out of seeing it here?

Reply:
@schustmi @AlexejPenner I removed the url methods. But Im a bit confused: Do you mean to leave out:

```
See the full code [here](https://github.com/zenml-io/zenml/blob/main/src/zenml/stack/flavor.py#L29).

See the full code [here](https://github.com/zenml-io/zenml/blob/main/src/zenml/stack/stack_component.py).
```

If yes then i would push back a bit... the link is important and it can change but the file wont change probably

Reply:
IMO I would keep the link, but remove all the code.

Reply:
Following this discussion I also took out `docs_url`. I didn't fully get which code section you proposed to remove @schustmi @AlexejPenner, `class StackComponent: ...` this one below? Or the code right here? Or both?